py-dask: add v2020.12.0

a-saitoh-fj opened the pull request:
Update to the latest version (2020.12.0)

a-saitoh-fj committed:
Update to the latest version (2020.12.0)

olesenm commented:
Duplicate of #20533

adamjstewart commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -120,6 +124,7 @@ class PyDask(PythonPackage):
+    depends_on('py-bokeh@1.0.0:1.999,2.0.1:', type=('build', 'run'), when='@2020.12.0: +diagnostics')
 
     # Requirements for dask.delayed
     depends_on('py-cloudpickle@0.2.1:', type=('build', 'run'), when='@2,7.0: +delayed')
```
It seems like the `+yaml` variant is no longer an option, `py-pyyaml` is always required now

adamjstewart reviewed (changes requested):
@skosukhin can you review? I'm guessing we'll want to remove the `yaml` variant, but I'll let you decide what you want to do.

a-saitoh-fj committed:
py-dask  modified:package.py

a-saitoh-fj commented:
py-dask
Fixed the problem that was pointed out.

- Removed yaml variant from default options (default=False)
- Added +yaml variant conflict condition (2.17.1 or later)
- Set py-pyyaml ​​to always depend (2.17.1 or later)

a-saitoh-fj commented:
@adamjstewart 
Is there anything wrong with this fix? If there is no problem, I hope it will be merged.

skosukhin commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -27,11 +28,11 @@ class PyDask(PythonPackage):
     variant('diagnostics', default=False, description='Install requirements for dask.diagnostics')
     variant('delayed',     default=True, description='Install requirements for dask.delayed (dask.imperative)')
-    variant('yaml',        default=True, description='Ensure support for YAML configuration files')
+    variant('yaml',        default=False, description='Ensure support for YAML configuration files')
```
I think we should keep the default value set to `True`.

skosukhin commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -130,7 +135,8 @@ class PyDask(PythonPackage):
     depends_on('py-toolz@0.8.2:',       type=('build', 'run'), when='@2.13.0: +delayed')
 
     # Support for YAML configuration files
-    depends_on('py-pyyaml',             type=('build', 'run'), when='+yaml')
```
If there are no objections to what I wrote regarding the conflict statements, we don't need to change anything here.

skosukhin commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -120,6 +124,7 @@ class PyDask(PythonPackage):
     # Requirements for dask.diagnostics
     depends_on('py-bokeh',              type=('build', 'run'), when='+diagnostics')
     depends_on('py-bokeh@1.0.0:',       type=('build', 'run'), when='@2.0.0: +diagnostics')
+    depends_on('py-bokeh@1.0.0:1.999,2.0.1:', type=('build', 'run'), when='@2020.12.0: +diagnostics')
```
If I'm not mistaken, the developers introduced the constraint starting version `2.26.0`:
```python
depends_on('py-bokeh@1.0.0:1.999,2.0.1:', type=('build', 'run'), when='@2.26.0: +diagnostics')
```

skosukhin commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -27,11 +28,11 @@ class PyDask(PythonPackage):
     conflicts('+distributed', when='@:0.4.0,0.7.6:0.8.1')
     conflicts('+diagnostics', when='@:0.5.0')
-    conflicts('+yaml', when='@:0.17.5')
+    conflicts('+yaml', when='@:0.17.5,2.17.1:')
```
The description of the `yaml` variant says "Ensure support for YAML configuration files". The reason it is formulated that way is that we don't really have full control over the feature: it is enabled whenever `py-pyyaml` is available (e.g. as a transitive dependency via `py-distributed`). So, `+yaml` basically says "make sure `py-pyyaml` is installed", which translates to `depends_on('py-pyyaml', when='+yaml')`. I admit that the meaning of `~yaml` is unclear from the description but, to my understanding, the easiest interpretation is "disable the direct dependency on `py-pyyaml`". Since the developers decided to make the dependency on `py-pyyaml` non-optional starting version `2.17.1`, the conflict statements should look like:
```python
conflicts('+yaml', when='@:0.17.5')
conflicts('~yaml', when='@2.17.1:')
```

skosukhin reviewed (commented):
@a-saitoh-fj sorry, I should have reviewed this much earlier

a-saitoh-fj committed:
py-dask modified:package.py (2nd time)

a-saitoh-fj commented:
@skosukhin
Thank you for your review.
Fixed as follows
- Changed yaml variant default to True
- As pointed out, I always enabled the yaml variant and set the ~yaml conflicts condition
- As pointed out, changed the dependency (version) of bokeh

skosukhin commented on var/spack/repos/builtin/packages/py-dask/package.py:
```diff
@@ -130,7 +136,7 @@ class PyDask(PythonPackage):
 
     # Support for YAML configuration files
-    depends_on('py-pyyaml',             type=('build', 'run'), when='+yaml')
+    depends_on('py-pyyaml',             type=('build', 'run'))
```
@a-saitoh-fj Shouldn't this be `when='+yaml'`? Otherwise, the `yaml` variant does nothing.

a-saitoh-fj committed:
py-dask modified:package.py (3rd time)

a-saitoh-fj commented:
@skosukhin
I'm sorry.
but the fix wasn't enough.
Corrected the pointed out part.

skosukhin commented:
@adamjstewart LGTM
